File: modules/db/src/blaze/db/impl/index/multi_version_id.clj
(ns blaze.db.impl.index.multi-version-id
  (:refer-clojure :exclude [conj])
  (:import
   [blaze.db.impl.index SingleVersionId MultiVersionId]))

(set! *warn-on-reflection* true)

(defn from-single-version-id
  "Creates a `MultiVersionId` from `single-version-id`."
  [single-version-id]
  (MultiVersionId/fromSingleVersionId ^SingleVersionId single-version-id))

(defn id
  "Returns the id of `multi-version-id`."
  [multi-version-id]
  (.id ^MultiVersionId multi-version-id))

(defn matches-hash?
  "Tests whether `multi-version-id` contains the version of `hash`."
  [multi-version-id hash]
  (.matchesHash ^MultiVersionId multi-version-id hash))

(defn conj
  "Returns a new `MultiVersionId` with the version of `single-version-id` added
  to that of `multi-version-id`."
  [multi-version-id single-version-id]
  (.conj ^MultiVersionId multi-version-id single-version-id))

File: modules/db/java/blaze/db/impl/index/MultiVersionId.java
package blaze.db.impl.index;

import blaze.ByteString;
import blaze.fhir.Hash;

import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public final class MultiVersionId {

    private final ByteString id;
    private final int[] hashPrefixes;

    private MultiVersionId(ByteString id, int[] hashPrefixes) {
        this.id = id;
        this.hashPrefixes = hashPrefixes;
    }

    public static MultiVersionId fromSingleVersionId(SingleVersionId singleVersionId) {
        return new MultiVersionId(singleVersionId.id(), new int[]{singleVersionId.hashPrefix()});
    }

    public ByteString id() {
        return id;
    }

    public boolean matchesHash(Hash hash) {
        return Arrays.binarySearch(hashPrefixes, hash.prefix()) >= 0;
    }

    public MultiVersionId conj(SingleVersionId singleVersionId) {
        int hashPrefix = singleVersionId.hashPrefix();
        int index = Arrays.binarySearch(hashPrefixes, hashPrefix);
        if (index >= 0) {
            return this;
        } else {
            index = -(index + 1);
            var newHashPrefixes = new int[hashPrefixes.length + 1];
            System.arraycopy(hashPrefixes, 0, newHashPrefixes, 0, hashPrefixes.length);
            System.arraycopy(hashPrefixes, index, newHashPrefixes, index + 1, hashPrefixes.length - index);
            newHashPrefixes[index] = hashPrefix;
            return new MultiVersionId(id, newHashPrefixes);
        }
    }

    @Override
    public String toString() {
        return "MultiVersionId{" +
                "id=" + id +
                ", hashPrefixes=[" + IntStream.of(hashPrefixes).mapToObj(h -> "0x" + Integer.toHexString(h)).collect(Collectors.joining(",")) +
                "]}";
    }
}
